### PR TITLE
Merge feature branch features/editorconfig-generated-code into master

### LINKED ISSUE #3705: Exclude files or folder from running through the analyzers through a configuration file.
Hi,

we have a lot of generated files that trigger warnings, errors when analyzers run them.
since the file name of the generated files do not follow the common naming conventions for generated code roslyn fires the analyzers away. As a result there are ten of thousands of warnings in the solution.
Could we get a exclude file from analyzing logic that:
- Exclude a file by adding it to a exclusion list
- Exclude a file by adding it to a exclusion list
- Annotate classes with a special Attribute that exclude the class
- exclude certain file name extensions ( like js files generated by typescript)
- Combination of the above like exclude all js files that live under the folder x including all sub folders

Most of the rules come straight from Resharper's exclusion file manager

Thanks

### PR DESCRIPTION
**Pull request by mavasani**
Fixes #3705

### CODE REVIEW

**Comment by mavasani** [comment]
@dotnet/roslyn-compiler - does PR merging a feature branch to master need 2 compiler signoffs?

File: src/Compilers/CSharp/Portable/PublicAPI.Unshipped.txt [hunk 1]
```diff
@@ -3,6 +3,11 @@
...
+*REMOVED*static Microsoft.CodeAnalysis.CSharp.CSharpSyntaxTree.ParseText(string text, Microsoft.CodeAnalysis.CSharp.CSharpParseOptions options = null, string path = "", System.Text.Encoding encoding = null, System.Collections.Immutable.ImmutableDictionary<string, Microsoft.CodeAnalysis.ReportDiagnostic> diagnosticOptions = null, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) -> Microsoft.CodeAnalysis.SyntaxTree
+*REMOVED*static Microsoft.CodeAnalysis.CSharp.SyntaxFactory.ParseSyntaxTree(Microsoft.CodeAnalysis.Text.SourceText text, Microsoft.CodeAnalysis.ParseOptions options = null, string path = "", System.Collections.Immutable.ImmutableDictionary<string, Microsoft.CodeAnalysis.ReportDiagnostic> diagnosticOptions = null, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) -> Microsoft.CodeAnalysis.SyntaxTree
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/PublicAPI.Unshipped.txt]: It looks like this API is missing a backcompat overload.

**Comment by 333fred** [comment]
>         public static SyntaxTree ParseSyntaxTree(

This API, for reference from my comment in the public API file

---
Refers to: src/Compilers/CSharp/Portable/Syntax/SyntaxFactory.cs:1569 in efb6a23. [](commit_id = efb6a23b69f6c893348980bb5e6580d8a2503e7b, deletion_comment = False)

File: src/Compilers/CSharp/Portable/PublicAPI.Unshipped.txt (hunk 1 above)
Comment (mavasani) [reply src/Compilers/CSharp/Portable/PublicAPI.Unshipped.txt]: This API is the back compat overload. It has the *REMOVED* prefix. Am I missing something?

**Comment by mavasani** [comment]
@333fred https://github.com/dotnet/roslyn/pull/40456/files#diff-7db012bc0e383f995338ed181e5cf399R2607-R2628 has the back compat overloads at the bottom of the same file. Did you expect something else?

File: src/Compilers/CSharp/Portable/PublicAPI.Unshipped.txt (hunk 1 above)
Comment (mavasani) [reply src/Compilers/CSharp/Portable/PublicAPI.Unshipped.txt]: Ah, I see the overload which takes `SourceText` as first parameter does not have the back compat overload. Thanks for the catch. I will push an additional commit to this PR to fix it.

Comment (mavasani) [reply src/Compilers/CSharp/Portable/PublicAPI.Unshipped.txt]: Created a separate PR https://github.com/dotnet/roslyn/pull/40504 as this feature branch is protected from direct commits.

**Comment by Nitin5acc** [comment]
Hi,

Are they any examples on how to use it? We do not want code analyzer to run on certain files in a solution.

**Pull request merged**